app/src/main/java/com/github/geohunt/app/ui/components/challenge/ChallengeViewModel.kt

```diff
@@ -0,0 +1,125 @@
+package com.github.geohunt.app.ui.components.challenge
+
+import android.app.Application
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import androidx.lifecycle.viewmodel.initializer
+import androidx.lifecycle.viewmodel.viewModelFactory
+import com.github.geohunt.app.data.repository.*
+import com.github.geohunt.app.model.Challenge
+import com.github.geohunt.app.model.Claim
+import com.github.geohunt.app.model.User
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+class ChallengeViewModel(
+    private val challengeRepository: ChallengeRepositoryInterface,
+    private val claimRepository: ClaimRepositoryInterface,
+    private val userRepository: UserRepositoryInterface,
+    private val authRepository: AuthRepositoryInterface,
+    private val followRepository: FollowRepositoryInterface,
+    private val activeHuntsRepository: ActiveHuntsRepositoryInterface,
+) : ViewModel() {
+
+    data class State(
+        val challenge: Challenge,
+        val author: User,
+        val claims: List<Claim>,
+        val isSelf: Boolean,
+        val authorScore: Long,
+        val doesFollow: Flow<Boolean>,
+        val doesHunt: Flow<Boolean>,
+        val alreadyClaimed: Boolean
+    )
+
+    private val state_ : MutableStateFlow<State?> = MutableStateFlow(null)
+    val state : StateFlow<State?> = state_
+
+    fun follow() {
+        require(state.value != null)
+        viewModelScope.launch {
+            followRepository.follow(state.value!!.author)
+        }
+    }
+
+    fun unfollow() {
+        require(state.value != null)
+        viewModelScope.launch {
+            followRepository.unfollow(state.value!!.author)
+        }
+    }
+
+    fun joinHunt() {
+        require(state.value != null)
+        viewModelScope.launch {
+            activeHuntsRepository.joinHunt(state.value!!.challenge)
+        }
+    }
+
+    fun leaveHunt() {
+        require(state.value != null)
+        viewModelScope.launch {
+            activeHuntsRepository.leaveHunt(state.value!!.challenge)
+        }
+    }
+
+    fun retrieveUser(uid: String) : StateFlow<User?> {
+        val mutableStateFlow = MutableStateFlow<User?>(null)
+
+        viewModelScope.launch {
+            mutableStateFlow.value = userRepository.getUser(uid)
+        }
+
+        return mutableStateFlow

```

Comment:
Make the `MutableStateFlow` read-only by using `asStateFlow`:
```suggestion
        return mutableStateFlow.asStateFlow()

```

Reply:
No need because I am already making the conversion. The return type of the function is `StateFlow<User?>` which is indeed read only

Reply:
Well I guess it prevent the user from casting it back so i'll do it

Reply:
Yeah, that was the main idea, just to be safe 👍